maven-core/src/test/java/org/apache/maven/graph/ProjectSelectorTest.java

```diff
@@ -0,0 +1,221 @@
+                () -> sut.getRequiredProjectsBySelectors( mavenExecutionRequest, listOfProjects, selectors ) );
+        assertThat( exception.getMessage(), containsString( "Could not find" ) );
+        assertThat( exception.getMessage(), containsString( ":required" ) );
+    }
+
+    @Test
+    void getRequiredProjectsBySelectorsReturnsProject() throws MavenExecutionException
+    {
+        final HashSet<String> selectors = new HashSet<>();
+        selectors.add( ":maven-core" );
+
+        final MavenProject mavenProject = getMavenProject( "maven-core" );
+        final List<MavenProject> listOfProjects = Collections.singletonList( mavenProject );
+
+        final Set<MavenProject> requiredProjectsBySelectors =
+                sut.getRequiredProjectsBySelectors( mavenExecutionRequest, listOfProjects, selectors );
+
+        assertThat( requiredProjectsBySelectors.size(), is( 1 ) );
+        assertThat( requiredProjectsBySelectors, contains( mavenProject ) );
+    }
+
+    @Test
+    void getRequiredProjectsBySelectorsReturnsProjectWithChildProjects() throws MavenExecutionException
+    {
+        when( mavenExecutionRequest.isRecursive() ).thenReturn( true );
+
+        final HashSet<String> selectors = new HashSet<>();
+        selectors.add( ":maven-core" );
+
+        final MavenProject mavenProject = getMavenProject( "maven-core" );
+        final MavenProject child = getMavenProject( "maven-core-child" );
+        mavenProject.setCollectedProjects( Collections.singletonList( child ) );
+        final List<MavenProject> listOfProjects = Collections.singletonList( mavenProject );
+
+        final Set<MavenProject> requiredProjectsBySelectors =
+                sut.getRequiredProjectsBySelectors( mavenExecutionRequest, listOfProjects, selectors );
+
+        assertThat( requiredProjectsBySelectors.size(), is( 2 ) );
+        assertThat( requiredProjectsBySelectors, contains( mavenProject, child ) );
+    }
+
+    @Test
+    void getOptionalProjectsBySelectorsReturnsProjectWithChildProjects()
+    {
+        when( mavenExecutionRequest.isRecursive() ).thenReturn( true );
+
+        final HashSet<String> selectors = new HashSet<>();
+        selectors.add( ":maven-core" );
+
+        final MavenProject mavenProject = getMavenProject( "maven-core" );
+        final MavenProject child = getMavenProject( "maven-core-child" );
+        mavenProject.setCollectedProjects( Collections.singletonList( child ) );
+        final List<MavenProject> listOfProjects = Collections.singletonList( mavenProject );
+
+        final Set<MavenProject> optionalProjectsBySelectors =
+                sut.getOptionalProjectsBySelectors( mavenExecutionRequest, listOfProjects, selectors );
+
+        assertThat( optionalProjectsBySelectors.size(), is( 2 ) );
+        assertThat( optionalProjectsBySelectors, contains( mavenProject, child ) );
+    }
+
+    private MavenProject getMavenProject( String artifactId )
```

Comment:
I'd suggest naming this method `createMavenProject` so it's clear that this creates an object that you will use only in the test. Reading the first few test methods confused me a bit...

Reply:
Fair point, changed it!